xcm/procedural/src/multilocation.rs

```diff
@@ -0,0 +1,188 @@
+	}
+
+	let from_tuples = generate_conversion_from_tuples();
+	let from_v0 = generate_conversion_from_v0();
+
+	Ok(quote! {
+		#from_tuples
+		#from_v0
+	})
+}
+
+fn generate_conversion_from_tuples() -> TokenStream {
+	let mut from_tuples = TokenStream::new();
+
+	for num_junctions in 0..8usize {
+		let junctions = (0..=num_junctions).map(|_| format_ident!("Junction")).collect::<Vec<_>>();
+		let idents = (0..=num_junctions).map(|i| format_ident!("j{}", i)).collect::<Vec<_>>();
+		let variant = &format_ident!("X{}", num_junctions + 1);
+		let array_size = num_junctions + 1;
+
+		let from_tuple = quote! {
+			impl From<( #(#junctions,)* )> for MultiLocation {
+				fn from( ( #(#idents,)* ): ( #(#junctions,)* ) ) -> Self {
+					MultiLocation { parents: 0, interior: Junctions::#variant( #(#idents),* ) }
+				}
+			}
+
+			impl From<(u8, #(#junctions),*)> for MultiLocation {
+				fn from( ( parents, #(#idents),* ): (u8, #(#junctions),* ) ) -> Self {
+					MultiLocation { parents, interior: Junctions::#variant( #(#idents),* ) }
+				}
+			}
+
+			impl From<[Junction; #array_size]> for MultiLocation {
+				fn from(j: [Junction; #array_size]) -> Self {
+					let [#(#idents),*] = j;
+					MultiLocation { parents: 0, interior: Junctions::#variant( #(#idents),* ) }
+				}
+			}
+		};
+
+		// Support up to 8 Parents in a tuple
+		for cur_parents in 1..=8u8 {
+			let parents = (0..cur_parents).map(|_| format_ident!("Parent")).collect::<Vec<_>>();
+			let underscores =
+				(0..cur_parents).map(|_| Token![_](Span::call_site())).collect::<Vec<_>>();
+
+			from_tuples.extend(quote! {
+				impl From<( #(#parents,)* #(#junctions),* )> for MultiLocation {
+					fn from( (#(#underscores,)* #(#idents),*): ( #(#parents,)* #(#junctions),* ) ) -> Self {
+						MultiLocation { parents: #cur_parents, interior: Junctions::#variant( #(#idents),* ) }
+					}
+				}
+			});
+		}
+
+		from_tuples.extend(from_tuple);
+	}
+
+	// Support up to 8 Parents in a tuple
+	for cur_parents in 1..=8u8 {
+		let parents = (0..cur_parents).map(|_| format_ident!("Parent")).collect::<Vec<_>>();
+		let underscores =
+			(0..cur_parents).map(|_| Token![_](Span::call_site())).collect::<Vec<_>>();
+
+		from_tuples.extend(quote! {
+			impl From<( #(#parents,)* Junctions )> for MultiLocation {
+				fn from( (#(#underscores,)* junctions): ( #(#parents,)* Junctions ) ) -> Self {
+					MultiLocation { parents: #cur_parents, interior: junctions }
+				}
+			}
+		});
+	}
+
+	quote! {
+		impl From<Junctions> for MultiLocation {
+			fn from(junctions: Junctions) -> Self {
+				MultiLocation { parents: 0, interior: junctions }
+			}
+		}
+
+		impl From<(u8, Junctions)> for MultiLocation {
+			fn from((parents, interior): (u8, Junctions)) -> Self {
+				MultiLocation { parents, interior }
+			}
+		}
+
+		impl From<()> for MultiLocation {
+			fn from(_: ()) -> Self {
+				MultiLocation { parents: 0, interior: Junctions::Here }
+			}
+		}
```

Comment:
I'm not sure I'm a fan of this.
It allows for `let multi: MultiLocation = ().into();` correct?